java/vector/src/test/java/org/apache/arrow/vector/TestCopyFrom.java

```diff
@@ -1069,4 +1070,29 @@ public void testCopyFromWithNulls14() {
       }
     }
   }
+
+  @Test //https://issues.apache.org/jira/browse/ARROW-7837
+  public void testCopySafeArrow7837() {
+    // this test exposes a bug in `handleSafe` where
+    // it reads a stale index and as a result missed a required resize of the value vector.
+    try (VarCharVector vc1 = new VarCharVector("vc1", allocator);
+         VarCharVector vc2 = new VarCharVector("vc2", allocator);
+    ) {
+      vc2.setInitialCapacity(20, 0.5);
+
+      vc1.setSafe(0, "1234567890".getBytes(Charset.forName("utf-8")));
+      assertFalse(vc1.isNull(0));
+      assertEquals(vc1.getObject(0).toString(), "1234567890");
+
+      vc2.copyFromSafe(0, 0, vc1);
+      assertFalse(vc2.isNull(0));
+      assertEquals(vc2.getObject(0).toString(), "1234567890");
+
```

Comment:
could you assert the intermediate elements are null? Or is this handled eslewhere.